### PR TITLE
Technical/aa 1101 accommodation mapper refactoring

### PR DESCRIPTION
**Pull request by hasmikharutyunyan**

### CODE REVIEW

File: HappyTravel.Nakijin.Updater/Infrastructure/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -126,12 +129,12 @@ public static IServiceCollection AddServices(this IServiceCollection services)
...
                         .Except(new Suppliers[] {Suppliers.Unknown}).ToList()
                     : suppliers.Split(';').Select(s => s.Trim()).Select(Enum.Parse<Suppliers>).ToList();
-            });                
+            });                          
 
             services.Configure<MappingOptions>(o =>
             {
                 var mappingMode = EnvironmentVariableHelper.Get("Workers:MappingOptions:MappingMode", configuration);
-                o.Mode = string.IsNullOrWhiteSpace(mappingMode)
+                 o.Mode = string.IsNullOrWhiteSpace(mappingMode)
```
Comment (kirillta) [inline HappyTravel.Nakijin.Updater/Infrastructure/ServiceCollectionExtensions.cs:137]: An extra space here

File: HappyTravel.Nakijin.Updater/Infrastructure/ServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -114,8 +117,8 @@ public static IServiceCollection AddServices(this IServiceCollection services)
             {
                 var workersToRun = EnvironmentVariableHelper.Get("Workers:WorkersToRun", configuration);
                 var suppliers = EnvironmentVariableHelper.Get("Workers:Suppliers", configuration);
-   
-                o.WorkersToRun = string.IsNullOrWhiteSpace(workersToRun)
+
+                 o.WorkersToRun = string.IsNullOrWhiteSpace(workersToRun)
```
Comment (kirillta) [inline HappyTravel.Nakijin.Updater/Infrastructure/ServiceCollectionExtensions.cs:121]: An extra space here

File: HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationValidator.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    public (bool IsValid, DeactivationReasons DeactivationReason) Validate(Suppliers supplier, MultilingualAccommodation accommodation)
+    {
+        if (accommodation.Location.Locality == null)
+        {
+            _logger.LogNotValidLocalityOfAccommodation(supplier.ToString(), accommodation.Location.Locality?.En);
+            return (false, DeactivationReasons.EmptyOrNotExistingLocality);
+        }
+
+        if (!accommodation.IsActive)
```
Comment (kirillta) [inline HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationValidator.cs:26]: Shouldn't this one be the first?

File: HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationDataStorageBySupplierAndCountry.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+
+    private async Task InitializeActiveUncertainMatches(Suppliers supplier, string countryCode, CancellationToken cancellationToken)
+        => _activeUncertainMatches =
+            await _accommodationMapperDataRetrieveService.GetActiveCountryUncertainMatchesBySupplier(countryCode, supplier, cancellationToken);
+
+
+    public ReadOnlyDictionary<int, (int SourceHtId, int HtIdToMatch)> ActiveUncertainMatches => new(_activeUncertainMatches);
+
```
Comment (kirillta) [inline HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationDataStorageBySupplierAndCountry.cs:65]: Do we need empty lines here?

File: HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationsBatchChangeCollector.cs [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+    {
+        int matchedHtId = 0;
+        if (_accommodationDataStorage.SupplierActiveAccommodations.TryGetValue(accommodation.SupplierCode, out var existing))
+        {
+            matchedHtId = existing.HtId;
+            if (IsExistUncertainMatch(matchedHtId, existingHtId))
+                return;
+        }
+        else if (_accommodationDataStorage.SupplierInvalidAccommodations.TryGetValue(accommodation.SupplierCode,
+                     out var existingNotActive))
```
Comment (kirillta) [inline HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationsBatchChangeCollector.cs:64]: Hmm

File: HappyTravel.Nakijin.Data/Migrations/20220128125135_ChangeDeactivationReasonToNewValues.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            migrationBuilder.Sql(@"UPDATE ""Accommodations""
+            SET ""DeactivationReason"" = 16
+            WHERE ""DeactivationReason"" = 5");
+             
+            migrationBuilder.Sql(@"UPDATE ""Accommodations""
+            SET ""DeactivationReason"" = 32
+            WHERE ""DeactivationReason"" = 6");
+            
+            migrationBuilder.Sql(@"UPDATE ""Accommodations""
+            SET ""DeactivationReason"" = 7
```
Comment (AlexeyShok) [inline HappyTravel.Nakijin.Data/Migrations/20220128125135_ChangeDeactivationReasonToNewValues.cs:32]: DeactivationReason can't be 7

**Review by AlexeyShok — CHANGES_REQUESTED** [review]

File: HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationDataStorageBySupplierAndCountry.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+public class AccommodationDataStorageBySupplierAndCountry
+{
+    public AccommodationDataStorageBySupplierAndCountry(IAccommodationMapperDataRetrieveService accommodationMapperDataRetrieveService)
+    {
+        _accommodationMapperDataRetrieveService = accommodationMapperDataRetrieveService;
+    }
+
+
+    public async Task Initialize(Suppliers supplier, string countryCode, CancellationToken cancellationToken)
```
Comment (oleg-n-k) [inline HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationDataStorageBySupplierAndCountry.cs:21]: cancellation token isn't used below

File: HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationDataStorageBySupplierAndCountry.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        _supplierDeprecatedToActivateAccommodations.Clear();
+
+        await InitializeNotSupplierAccommodationsTree(supplier, countryCode, cancellationToken);
+        await InitializeSupplierAccommodations(supplier, countryCode, cancellationToken);
+        await InitializeActiveUncertainMatches(supplier, countryCode, cancellationToken);
+    }
+
+
+    private async Task InitializeNotSupplierAccommodationsTree(Suppliers supplier, string countryCode, CancellationToken cancellationToken)
+        => NotSupplierAccommodationsTree = await _accommodationMapperDataRetrieveService.GetCountryAccommodationsTree(countryCode, supplier);
```
Comment (oleg-n-k) [inline HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationDataStorageBySupplierAndCountry.cs:35]: Cancellation token can be used here

File: HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationDataStorageBySupplierAndCountry.cs [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+
+    private async Task InitializeSupplierAccommodations(Suppliers supplier, string countryCode, CancellationToken cancellationToken)
+    {
+        var accommodationsOfSupplier = await _accommodationMapperDataRetrieveService.GeCountryAccommodationBySupplier(countryCode, supplier);
```
Comment (oleg-n-k) [inline HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationDataStorageBySupplierAndCountry.cs:40]: Ge**t**CountryAccommodationBySupplier
Cancellation token can be used here

File: HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationDataStorageBySupplierAndCountry.cs [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+            await _accommodationMapperDataRetrieveService.GetActiveCountryUncertainMatchesBySupplier(countryCode, supplier, cancellationToken);
+
+
+    public ReadOnlyDictionary<int, (int SourceHtId, int HtIdToMatch)> ActiveUncertainMatches => new(_activeUncertainMatches);
+    public ReadOnlyDictionary<string, SlimAccommodationData> SupplierActiveAccommodations => new(_supplierActiveAccommodations);
+    public ReadOnlyDictionary<string, SlimAccommodationData> SupplierDeprecatedToActivateAccommodations => new(_supplierDeprecatedToActivateAccommodations);
+    public ReadOnlyDictionary<string, SlimAccommodationData> SupplierInvalidAccommodations => new(_supplierInvalidAccommodations);
+    public STRtree<SlimAccommodationData> NotSupplierAccommodationsTree = new();
+    
+
```
Comment (oleg-n-k) [inline HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationDataStorageBySupplierAndCountry.cs:70]: extra line

**Review by AlexeyShok — APPROVED** [review]

**Review by oleg-n-k — APPROVED** [review]

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationDataCorrectionService.cs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        var (_, isFailure, accommodationId, error) = ManagementHelper.RetrieveAccommodationId(accommodationHtId);
+        if (isFailure)
+            return Result.Failure(error);
+
+        var dbAccommodation = await _context.Accommodations
+            .Where(ac => ac.IsActive && ac.Id == accommodationId)
+            .Include(ac => ac.Locality)
+            .Include(ac => ac.Country)
+            .Include(ac => ac.LocalityZone)
+            .SingleOrDefaultAsync();
```
Comment (kirillta) [inline HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationDataCorrectionService.cs:33]: Why do you need two exactly similar methods except this very line? You may reuse the second one here

File: HappyTravel.Nakijin.Core/Services/MultilingualDataHelper.cs [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+            return accommodation.Contacts;
+
+        var contacts = Sanitize(accommodation.Contacts);
+        _logger.LogInvalidContactInfo(accommodation.SupplierCode);
+
+        return contacts;
+    }
+
+
+    private bool IsSanitizingNeeded(ContactInfo contactInfo)
```
Comment (kirillta) [inline HappyTravel.Nakijin.Core/Services/MultilingualDataHelper.cs:141]: Do we need this check if in `Sanitize` method we check it again?

Comment (hasmikharutyunyan) [reply HappyTravel.Nakijin.Core/Services/MultilingualDataHelper.cs:141]: I think it is done for logging error if all contacts of supplier not valid

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationDataCorrectionService.cs (hunk 11 above)
Comment (hasmikharutyunyan) [reply HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationDataCorrectionService.cs:33]: the core logic of methods is done in one place, here just the query to the db is different, I think it is ok.

Comment (hasmikharutyunyan) [reply HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationDataCorrectionService.cs]: Changed

File: HappyTravel.Nakijin.Data/Extensions/NakijinContextExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,17 @@
+using System.Linq;
+using Microsoft.EntityFrameworkCore;
+
+namespace HappyTravel.Nakijin.Data.Extensions;
+
+public static class NakijinContextExtensions
+{
+    public static void DetachAll(this NakijinContext context)
+    {
+        var attachedEntities = context.ChangeTracker.Entries()
```
[the remark this answers is not in the document]
Comment (kirillta) [reply HappyTravel.Nakijin.Data/Extensions/NakijinContextExtensions.cs:10]: But its mock could. I speak from experience

File: HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationsBatchChangeCollector.cs [hunk 14]
```diff
@@ -95,7 +95,7 @@ public void RemoveSupplierCodeOrDeactivateNotValid(Contracts.MultilingualAccommo
         if (!_accommodationDataStorage.SupplierActiveAccommodations.TryGetValue(accommodation.SupplierCode, out var existingAccommodation)
             || existingAccommodation.SupplierAccommodationCodes.Count == 1)
         {
-            DeactivateOrAddNotActive(accommodation.SupplierCode, deactivationReason, accommodation);
+            ProcessDeactivation(accommodation.SupplierCode, deactivationReason, accommodation);
```
Comment (oleg-n-k) [inline HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationsBatchChangeCollector.cs:98]: Just Deactivate maybe?

File: HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationsBatchChangeCollector.cs [hunk 15]
```diff
@@ -229,14 +222,50 @@ void FindAndDeactivateUncertainMatch(int firstHtId, int secondHtId)
                 entry.Property(um => um.Modified).IsModified = true;
             }
         }
+
+
+        void Merge(SlimAccommodationData existing)
+        {
+            ProcessDeactivation(accommodation.SupplierCode, DeactivationReasons.MatchingWithOther, accommodation);
```
Comment (oleg-n-k) [inline HappyTravel.Nakijin.Updater/Workers/AccommodationMapping/AccommodationsBatchChangeCollector.cs:229]: How about just using the name "Deactivate"?

File: HappyTravel.Nakijin.Data/Extensions/NakijinContextExtensions.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (hasmikharutyunyan) [reply HappyTravel.Nakijin.Data/Extensions/NakijinContextExtensions.cs]: Done

**Review by kirillta — APPROVED** [review]

**Pull request merged**